### DISCUSSION TITLE
How can I know who triggers on_member_update?

### DISCUSSION THREAD

**Discussion by riccardolunardi** [Rapptz/discord.py#6182] [Q&A] [answered]
In my server everyone can change other people's nicknames.
`on_member_update(before, after)` give the ability to know how the user changed _after_ the update, but how can I know **who** made the update?

I tried looking on the intenet but didn't find anything helpful :(

**Comment by ghost** [comment, marked as answer]
You can iterate through your server's audit log

Basically, since `before` and `after` are [`Member`](https://discordpy.readthedocs.io/en/latest/api.html?highlight=on_member_update#member) objects, you can get the guild from them:

```py
@commands.Cog.listener()
async def on_member_update(self, before, after):
  async for entry in after.guild.audit_logs(limit=10):
      print('{0.user} did {0.action} to {0.target}'.format(entry))
```

I don't know if this is the best way to do that, but it works. Another way is to get the guild from a specific ID:
```py
@commands.Cog.listener()
async def on_member_update(self, before, after):
  guild = self.bot.get_guild(123456789)
  async for entry in guild.audit_logs(limit=10):
      print('{0.user} did {0.action} to {0.target}'.format(entry))
```

Check out these references for more information: [`on_member_update`](https://discordpy.readthedocs.io/en/latest/api.html?highlight=on_member_update#discord.on_member_update), [`audit_logs`](https://discordpy.readthedocs.io/en/latest/api.html#discord.Guild.audit_logs)

**Reply by riccardolunardi** [reply]
Thanks, that's what I thought I should do, too. I wondered if there was another way, but this look like the one to go.

**Reply by EvokerKing** [reply]
You should mark it as an answer.
